src/basilisp/core.lpy

```diff
@@ -7085,3 +7085,122 @@
      true
      (catch queue/Full _ false))))
+
+;;;;;;;;;;;
+;; Proxy ;;
+;;;;;;;;;;;
+
+(defmacro proxy [class-and-interfaces args & fs]
+  "Expands to code that creates an instance of a class that implements
+  the Basilisp interfaces in ``class-and-interfaces``. The methods
+  implementations are provided in ``fs``, while the ``args`` are
+  passed to superclass constructor of the instance. If no class is
+  specified in ``class-and-interfaces``, the proxy instance inherits
+  from `python/object`.
+
+  ``fs`` consists of class/interface method definitions and/or
+  multiarity interface method definitions
+
+  (method-name1 [args...] body)
+  (method-name2 ([args1...] body1)
+                ([args2...] body2) ...)
+  ...
+
+  A `this` argument is implicitly avaialble in method bodies, refering
+  to the proxy instance.
+
+  The single arity version can also accept python keyword args using
+  the meta :kwargs option, as follows
+
+  ^{:kwargs :collect} (method-name [args... {:as kwargs}])
+
+  The interfaces in ``class-and-interfaces`` can be regular classes or
+  abstract classes whose methods can be overriden in ``fs``.
+
+  Moreover, if a method in ``fs`` is not found in any of the methods
+  in ``class-and-interfaces``, it is added as a new method in the
+  proxy instance."
+  (let [class-nm (str (gensym (str "proxy--" (first class-and-interfaces))))
+        fname-fn-pairs
+        (for [f fs]
+          (if (list? (second f))
+            ;; multi-arity
+            ;;
+            ;; Create a single variadic fn that will switch to the
+            ;; correct body based on the number of arguments passed
+            ;; in.
+            (let [fname-fn-pair
+                  (let [vargs (gensym "args")
+                        fname (first f)
+                        ;; sort the different arities to ensure that
+                        ;; variadic arity comes last, courtesy of the
+                        ;; additional `&` symbol.
+                        fargs (sort-by count (rest f))
+                        fargs-last (last fargs)
+                        variadic-pos (try
+                                       (.index (first fargs-last) '&)
+                                       (catch python/ValueError _
+                                         nil))]
+                    [(str fname)
+                     `(fn [& ~vargs]
+                        ;; create a case form to switch to the right
+                        ;; body according to the number of arguments
+                        ;; passed to the fn.
+                        ;;
+                        ;; (case (count fn-args)
+                        ;;   n1 (do body1)
+                        ;;   n2 (do body2)
+                        ;;   (if-not (variadic? fn-args)
+                        ;;     (throw exception)
+                        ;;     (do variadic-body))
+                        ~(apply list 'case (list 'count vargs)
```

Comment:
```suggestion
                        ~(apply list `case `(count ~vargs)
```

Does this work? I don't think we want to use normal quotes for names that will refer to known vars.

Reply:
It does, thanks, updated.